temporal-server/authorizer/claimer.go

```diff
@@ -0,0 +1,245 @@
+	ctx := context.Background()
+	namespaceAccess, err := c.NamespaceAccessProvider.GetNamespaceAccessInformation(ctx, email)
+	if err != nil {
+		return nil, c.generateError(fmt.Sprintf("error reading namespace access: %v \n", err))
+	}
+
+	openAccessNamespaces := strings.Split(c.OpenAccessNamespaces, ",")
+	for _, ns := range openAccessNamespaces {
+		claims.Namespaces[ns] = authorization.RoleWriter
+	}
+
+	adminGroupsSlice := strings.Split(c.AdminGroups, ",")
+	for _, ns := range namespaceAccess {
+		if ns.Namespace != "" {
+			if slices.Contains(adminGroupsSlice, ns.Namespace) && ns.Relation == "admin" {
+				claims.System = authorization.RoleWriter
+			} else {
+				role, exists := roleMap[ns.Relation]
+				if exists {
+					claims.Namespaces[ns.Namespace] = role
+				}
+			}
+		}
+	}
+
+	if len(claims.Namespaces) > 0 {
+		claims.Namespaces[""] = authorization.RoleReader
+	}
+
+	return &claims, nil
+}
+
+// GetNamespaceAccessInformation returns a list of namespaces that a user with the given email
+// has access to along with the type of relation they have (One of "reader", "writer" or "admin").
+func (c *AuthClient) GetNamespaceAccessInformation(ctx context.Context, email string) ([]NamespaceAccess, error) {
+	groups, err := c.getUserGroups(ctx, email)
+	if err != nil {
+		return nil, err
+	}
+
+	var namespaceAccess []NamespaceAccess
+	for _, group := range groups {
+		continuationToken := ""
+
+		for {
+			tuples, continuationToken, err := c.OfgaClient.FindMatchingTuples(ctx, ofga.Tuple{
+				Object:   &ofga.Entity{Kind: "group", ID: fmt.Sprintf("%v#member", group)},
+				Relation: "",
+				Target:   &ofga.Entity{Kind: "namespace", ID: ""},
+			}, 100, continuationToken)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, tuple := range tuples {
+				namespaceAccess = append(namespaceAccess, NamespaceAccess{
+					Namespace: tuple.Tuple.Target.ID,
+					Relation:  tuple.Tuple.Relation.String(),
+				})
+			}
+
+			if continuationToken == "" {
+				break
+			}
+		}
+	}
+
+	return namespaceAccess, nil
+}
+
+// getUserGroups returns a list of groups that a user with the given email
+// is a member of in the OpenFGA store.
+func (c *AuthClient) getUserGroups(ctx context.Context, email string) ([]string, error) {
+	var groups []string
+	continuationToken := ""
+	for {
+		tuples, continuationToken, err := c.OfgaClient.FindMatchingTuples(ctx, ofga.Tuple{
+			Object:   &ofga.Entity{Kind: "user", ID: email},
+			Relation: "",
+			Target:   &ofga.Entity{Kind: "group", ID: ""},
+		}, 100, continuationToken)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, tuple := range tuples {
+			groups = append(groups, tuple.Tuple.Target.ID)
+		}
+
+		if continuationToken == "" {
+			break
+		}
+	}
+
+	return groups, nil
+}
+
+// generateError returns a new error and also logs it on the provided logger.
+func (c TokenClaimMapper) generateError(msg string) error {
+	if c.Logger != nil {
+		c.Logger.Error(msg)
+	}
+	return errors.New(msg)
+}
```

Comment:
This function seems to be conflating two things: building and error and logging it. And it's also a bit weird that this functions is a part of the `TokenClaimMapper`, since it's implementing a pattern, and not something inherent to `TokenClaimMapper`.

An alternative would be building be having an named return parameter in the functions, and then a defer function that checks that err and logs it if it's not nil. See this example in ua-contracts:
https://github.com/canonical/ua-contracts/blob/8801b8c9085c55607edce8be8ba74448e45a4d4c/internal/postgresstore/store.go#L2320-L2336

And how it gets invoked by others functions in the same file: https://github.com/canonical/ua-contracts/blob/8801b8c9085c55607edce8be8ba74448e45a4d4c/internal/postgresstore/store.go#L114-L119

By doing that, you are free to build errors using Go's native function for that: 
`fmt.Errorf`, and worry about logging and monitor separately.